SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp

```diff
@@ -0,0 +1,250 @@
+  static std::string get_description() {
+    return "atomic update with zero operands";
+  }
+
+  template <typename DataT, int NumElems, int NumElemsToChange,
+            typename ChangeElemFilterT, atomic_op ChosenOperator>
+  static void call_esimd_function(DataT *input_data, DataT *output_data,
+                                  size_t *const offsets,
+                                  size_t work_item_index) {
+    simd<DataT, NumElemsToChange> offset;
+    mask_type_t<NumElemsToChange> mask;
+
+    ChangeElemFilterT filter{};
+    for (size_t i = 0; i < NumElemsToChange; ++i) {
+      mask[i] = filter(i);
+      offset[i] = offsets[i] * sizeof(DataT);
+    }
+
+    auto values_before_update =
+        atomic_update<ChosenOperator, DataT, NumElemsToChange>(input_data,
+                                                               offset, mask);
+
+    unsigned int values_sum = 0;
+    for (size_t i = 0; i < NumElemsToChange; ++i) {
+      values_sum += values_before_update[i];
+    }
+
+    output_data[work_item_index] = values_sum;
+  }
+};
+
+// Provides expected value for updated memory locations.
+template <esimd::atomic_op Operator, typename T>
+T get_expected_value(T base_value, int number_of_iteractions) {
+  if constexpr (Operator == esimd::atomic_op::dec) {
+    return base_value - number_of_iteractions;
+  } else if constexpr (Operator == esimd::atomic_op::inc) {
+    return base_value + number_of_iteractions;
+  } else {
+    static_assert(Operator != Operator, "Unexpected  operator type.");
+  }
+}
+
+// The main test routine.
+// Using functor class to be able to iterate over the pre-defined data types.
+template <typename TestCaseT, typename DataT, typename NumElemsT,
+          typename NDRangeDimT, typename NumElemsToChangeT,
+          typename ChangeElemFilterT, typename ChosenOperatorT,
+          typename AlgorithmToChangeT>
+class run_test {
+  static constexpr int NumElems = NumElemsT::value;
+  static constexpr int NDRangeDim = NDRangeDimT::value;
+  static constexpr int NumElemsToChange = NumElemsToChangeT::value;
+  static constexpr atomic_op ChosenOperator = ChosenOperatorT::value;
+  static constexpr functions::offset AlgorithmToChange =
+      AlgorithmToChangeT::value;
+  using TestDescriptionT = functions::TestDescription<NumElems, TestCaseT>;
+
+public:
+  bool operator()(sycl::queue &queue, const std::string &data_type,
+                  const std::string &operator_type) {
+    static_assert(NDRangeDim <= 3 && NDRangeDim > 0);
+    static_assert(NumElemsToChange <= NumElems);
+    bool passed = true;
+    constexpr DataT base_value = 100;
+    constexpr DataT result_fill_value = 5;
+
+    auto init_values =
+        functions::get_init_values<NumElemsToChange, ChosenOperator, DataT>(
+            base_value);
+    shared_allocator<DataT> allocator(queue);
+    shared_vector<DataT> shared_init_values(init_values.begin(),
+                                            init_values.end(), allocator);
```

Comment:
I was sure you've told that you'd passed allocator to the factory function `get_init_values` itself in https://github.com/intel/llvm-test-suite/pull/939#discussion_r834023451, but I see no any. Is it some some local stash not pushed or do we misunderstand each other maybe?

Reply:
The `get_offsets` and `get_init_values` function has been renamed into `fill_*` and now they obtain a vector that will be filled, so intermediate variables like `init_values` and `offsets` are useless and they has been deleted
I've done it in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/f79221a335ddfdbd86622f5b4e1d73b1343acb24